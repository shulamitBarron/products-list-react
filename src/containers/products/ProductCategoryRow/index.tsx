import * as React from 'react';

interface Props {
	category: string;
}

const ProductCategoryRow: React.FC<Props> = (props: Props) => {
	const { category } = props;
	return (
		<tr>
			<th colSpan={3}>
				{category}
			</th>
		</tr>
	);
};

export default React.memo(ProductCategoryRow);
