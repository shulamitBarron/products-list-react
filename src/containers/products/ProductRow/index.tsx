import { Product } from 'actions/redux/product/interfaces';
import * as React from 'react';

interface Props {
	product: Product;
	selected: boolean;
	onProductSelected: (product: Product) => void;
}

const ProductRow: React.FC<Props> = (props: Props) => {
	const { product, selected, onProductSelected } = props;

	function onRowSelected(e: React.MouseEvent) {
		e.preventDefault();
		onProductSelected(product);
	}

	return (
		<tr onClick={onRowSelected} style={{ color: !product.isInStock ? 'red' : '', backgroundColor: selected ? '#9999ab' : '' }}>
			<td>{product.id}</td>
			<td>{product.name}</td>
			<td>{product.price}</td>
		</tr>
	);
};

export default React.memo(ProductRow);
