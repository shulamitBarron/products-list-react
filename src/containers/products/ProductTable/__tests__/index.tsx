import React from 'react';
import { shallow } from 'enzyme';
import { Table } from 'react-bootstrap';
import ProductTable from '../index';
import ProductCategoryRow from '../../ProductCategoryRow';
import ProductRow from '../../ProductRow';

const products = [
	{
		id: '5ff81abbe6ab17fd4601c5d8',
		isInStock: true,
		price: '2638.28',
		picture: 'http://placehold.it/32x32',
		name: 'Bray',
		group: 'sport',
		description: 'Minim enim mollit ea id dolor duis irure in.',
		registered: '2016-03-27 20:32:49'
	},
	{
		id: '5ff81abb0e750ff69443daf8',
		isInStock: true,
		price: '3554.61',
		picture: 'http://placehold.it/32x32',
		name: 'Hewitt',
		group: 'sport',
		description: 'Officia adipisicing velit anim deserunt officia est qui excepteur ullamco laborum.',
		registered: '2018-12-14 15:32:56'
	}
];

describe('ProductTable', () => {
	it('should render component', () => {
		const tree = shallow(<ProductTable products={products} onProductSelected={jest.fn()} translate={jest.fn()} selectedProductId="" />);

		expect(tree).toMatchSnapshot();
	});

	it('should render with correct childern', () => {
		const tree = shallow(<ProductTable products={products} onProductSelected={jest.fn()} translate={jest.fn()} selectedProductId="" />);

		expect(tree.find(Table).exists).toBeTruthy();

		expect(tree.find(ProductCategoryRow).length).toBe(1);
		expect(tree.find(ProductCategoryRow).first().props().category).toBe('sport');

		expect(tree.find(ProductRow).length).toBe(2);
		expect(tree.find(ProductRow).at(0).props().product).toBe(products[0]);
		expect(tree.find(ProductRow).at(1).props().product).toBe(products[1]);
	});

	it('should render empty table', () => {
		const tree = shallow(<ProductTable products={[]} onProductSelected={jest.fn()} translate={jest.fn()} selectedProductId="" />);

		expect(tree.find(Table).exists).toBeTruthy();

		expect(tree.find(ProductCategoryRow).length).toBe(0);
		expect(tree.find(ProductRow).length).toBe(0);
	});

	it('should call onProductSelected', () => {
		const onProductSelected = jest.fn();
		const tree = shallow(<ProductTable products={products} onProductSelected={onProductSelected} translate={jest.fn()} selectedProductId="" />);

		tree.find(ProductRow).first().props().onProductSelected(products[0]);
		expect(onProductSelected).toHaveBeenCalledWith(products[0]);
	});

	it('should send correct selected product', () => {
		const tree = shallow(<ProductTable products={products} onProductSelected={jest.fn()} translate={jest.fn()} selectedProductId="5ff81abbe6ab17fd4601c5d8" />);

		expect(tree.find(ProductRow).length).toBe(2);
		expect(tree.find(ProductRow).at(0).props().selected).toBe(true);
		expect(tree.find(ProductRow).at(1).props().selected).toBe(false);
	});
});
