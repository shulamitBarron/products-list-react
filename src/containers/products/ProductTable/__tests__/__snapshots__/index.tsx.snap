// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTable should render component 1`] = `
<ForwardRef>
  <thead>
    <tr>
      <th>
        #
      </th>
      <th />
      <th />
    </tr>
  </thead>
  <tbody>
    <ProductCategoryRow
      category="sport"
      key="sport"
    />
    <ProductRow
      key="5ff81abbe6ab17fd4601c5d8"
      onProductSelected={[MockFunction]}
      product={
        Object {
          "description": "Minim enim mollit ea id dolor duis irure in. Aliqua eiusmod sunt dolore laboris exercitation elit nostrud cupidatat et irure cupidatat sint aliquip labore. Excepteur cupidatat quis veniam labore elit. Aliqua quis ut sit sint occaecat eiusmod in enim ex ut est. Id officia id labore nostrud velit ipsum consequat ex est reprehenderit. Cupidatat sit sunt esse nulla officia incididunt eu deserunt amet qui dolore ea anim cillum.",
          "group": "sport",
          "id": "5ff81abbe6ab17fd4601c5d8",
          "isInStock": true,
          "name": "Bray",
          "picture": "http://placehold.it/32x32",
          "price": "2638.28",
          "registered": "2016-03-27 20:32:49",
        }
      }
      selected={false}
    />
    <ProductRow
      key="5ff81abb0e750ff69443daf8"
      onProductSelected={[MockFunction]}
      product={
        Object {
          "description": "Officia adipisicing velit anim deserunt officia est qui excepteur ullamco laborum. Aute laborum do do do veniam pariatur sint qui labore non fugiat occaecat. Est elit fugiat commodo tempor cupidatat duis culpa ipsum duis labore in et ut mollit. Tempor ipsum deserunt ad proident eu elit incididunt exercitation nisi et fugiat ut ut laborum. Ullamco voluptate reprehenderit qui laboris elit pariatur id cillum non pariatur ullamco consequat est. Consectetur occaecat mollit veniam reprehenderit.",
          "group": "sport",
          "id": "5ff81abb0e750ff69443daf8",
          "isInStock": true,
          "name": "Hewitt",
          "picture": "http://placehold.it/32x32",
          "price": "3554.61",
          "registered": "2018-12-14 15:32:56",
        }
      }
      selected={false}
    />
  </tbody>
</ForwardRef>
`;
